aptos-move/block-executor/src/proptest_types/types.rs

```diff
@@ -130,11 +166,24 @@ impl<K: Hash + Clone + Debug + Eq + PartialOrd + Ord> ModulePath for KeyType<K>
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""), PartialEq(bound = ""), Eq(bound = ""))]
 pub(crate) struct ValueType {
     /// Wrapping the types used for testing to add TransactionWrite trait implementation (below).
     bytes: Option<Bytes>,
     metadata: StateValueMetadataKind,
+    #[derivative(PartialEq = "ignore")]
+    write_op_kind: ExplicitSyncWrapper<WriteOpKind>,
```

Comment:
What's use of ExplicitSyncWrapper here?

Reply:
So, if WriteOpKind is deletion, instead of deleting the value, are we going to store the value and mark it as deletion in `write_op_kind`?

Reply:
ExplicitSyncWrapper is an implementation detail, we want to be as efficient inside mock transaction execution as possible, so avoiding mutexes (same txn is always executed single threaded - otherwise many other tests will break:)

The mock write ops need to have the correct kind, otherwise the block executor will now assert. So we replicate what the system does now, i.e. check existence - the system confirms and throws invariant error if it doesn't match, i.e. somehow Move-VM was trying to delete or modify smt that didn't exist, or create smt that already exists. Mock transaction execution instead adjusts the test op kinds so it fits and avoids the asserts. This is also a good test as if smt is wrong with existence checks or elsewhere the compatibility checks in block executor would fire.